### PR TITLE
alex/forgot reset password

### PR DESCRIPTION
**Pull request by larzhole**
## Description
Ability for non logged in users who forgot their password to request a password reset. For now this logs a generated token and in the future will email a link to the users' email address.

Ability for a logged in user to reset their password given the currentPassword, password, and passwordConfirmation.

Ability for a non-logged in user to reset their password given a token, userId, password, and passwordConfirmation.

Closes https://github.com/ericdowell/rerun/issues/3

**Issue(s):**
I was unable to generate the migration for the passwordResets table so I manually created the migration.
Error seems misleading
[2020-07-02T13_56_38_069Z-debug.log](https://github.com/ericdowell/rerun/files/4864956/2020-07-02T13_56_38_069Z-debug.log)

Now I'm not able to run migrations
Here is the error I'm getting

$ typeorm migration:run
Error during migration run:
/Developer/Work/rerun/src/entity/index.ts:1
export * from './PasswordReset'
^^^^^^

SyntaxError: Unexpected token 'export'
    at wrapSafe (internal/modules/cjs/loader.js:1054:16)
    at Module._compile (internal/modules/cjs/loader.js:1102:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
    at Module.load (internal/modules/cjs/loader.js:986:32)
    at Function.Module._load (internal/modules/cjs/loader.js:879:14)
    at Module.require (internal/modules/cjs/loader.js:1026:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Function.PlatformTools.load (/usr/local/lib/node_modules/typeorm/platform/PlatformTools.js:114:28)
    at /usr/local/lib/node_modules/typeorm/util/DirectoryExportedClassesLoader.js:39:69
    at Array.map (<anonymous>)

### CODE REVIEW

File: src/controllers/auth/resetPassword.ts [hunk 1]
```diff
@@ -1,26 +1,108 @@
...
 
+// Auth users reset password
+// /user/password/reset
 export const postUpdatePassword = async (req: Request, res: Response): Promise<Response> => {
-  await check('password', 'Password must be at least 4 characters long').isLength({ min: 8 }).run(req)
-  await check('confirmPassword', 'Passwords do not match').equals(req.body.password).run(req)
+  // TODO: make this into a middleware where we can apply it on a route-by-route basis.
+  /** @see: https://expressjs.com/en/guide/using-middleware.html **/
+  if (!req.user) {
+    return res.status(401).json({ error: 'not logged in.' })
```
Comment (larzhole) [inline src/controllers/auth/resetPassword.ts:14]: @ericdowell This is causing tests to fail. Do you know of a way to add a user to the request with supertest?

Comment (ericdowell) [reply src/controllers/auth/resetPassword.ts:14]: I didn’t find anything when I research, but I only spent minimal time doing that. If I can’t find something within superheat I was thinking about this and we can make a Magic Link login middleware which we’ll use for testing only.

We’d pass A JWT token and allow that request to authenticate the current request and/or session. 

I’ll look into Supertest some more before doing that.

File: src/migration/1593699407760-CreatePasswordResetsTable.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        columns: [
+          {
+            name: 'id',
+            type: 'bigint',
+            isPrimary: true,
+          },
+          {
+            name: 'userId',
+            type: 'bigint',
+          },
```
Comment (ericdowell) [inline src/migration/1593699407760-CreatePasswordResetsTable.ts:17]: @larzhole You'll need to use `npm run typeorm migration:run` as outlined in the [README.md](https://github.com/ericdowell/rerun/blob/f4cb8a6250d18144076b341de296cf9be7e03779/README.md#migrating-the-database) to get migrations to work.

Also make sure this `userId` is unsigned and unique as outlined in the Issue #3

File: src/migration/1593699407760-CreatePasswordResetsTable.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import { MigrationInterface, QueryRunner, Table, TableForeignKey } from 'typeorm'
+
+export class CreatePasswordResetsTable1593699407760 implements MigrationInterface {
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.createTable(
+      new Table({
+        name: 'passwordResets',
+        columns: [
+          {
+            name: 'id',
```
Comment (ericdowell) [inline src/migration/1593699407760-CreatePasswordResetsTable.ts:10]: Do we need a primary `id` for this table since `userId` is kind already handling that?

File: src/controllers/auth/resetPassword.ts [hunk 4]
```diff
@@ -1,26 +1,108 @@
...
+  let passwordReset: PasswordReset
+
+  try {
+    passwordReset = await PasswordReset.findOneOrFail({ where: { user } })
+  } catch (error) {
+    logger.error(error)
+    return res.status(404).json({ error: 'Token not found for the given email address' })
+  }
+
+  const createdAt = passwordReset.createdAt.toString()
```
Comment (ericdowell) [inline src/controllers/auth/resetPassword.ts:80]: @larzhole Why not just use `DateTime.fromISO()` here since the date is likely something like `2020-06-30T20:02:24.000Z`? See more about parsing with luxon [here](https://moment.github.io/luxon/docs/manual/parsing.html).

Does this work?
```suggestion
  const createdAt = DateTime.fromISO(passwordReset.createdAt)
```

File: src/auth/passwordBroker.ts [hunk 5]
```diff
@@ -1,5 +1,15 @@
 import { createHmac, randomBytes } from 'crypto'
+import { DateTime } from 'luxon'
+import { PasswordReset, User } from '../entity'
 
 export const createNewToken = (): string => {
   return createHmac('sha256', 'secret').update(randomBytes(40).toString('hex')).digest('hex')
 }
+
+export const createNewPasswordReset = async (user: User): Promise<PasswordReset> => {
```
Comment (ericdowell) [inline src/auth/passwordBroker.ts:9]: This should be passing the `token` into the function, which is why the `TODO` was specifically on the line after `const token`:
https://github.com/ericdowell/rerun/blob/55c5cf4cc64a3f08e904daedcd1393e983032ee9/src/controllers/auth/forgotPassword.ts#L24-L29

`token` should be created within context of the controller since we need to email that unhash value to the user to call the Reset Password endpoint.

File: src/controllers/auth/resetPassword.ts [hunk 6]
```diff
@@ -1,26 +1,108 @@
...
+    zone: 'America/Chicago',
+  }).toUTC()
+  const createdAtDateTimePlus60Minutes = createdAtDateTime.plus(Duration.fromObject({ minutes: 60 }))
+
+  if (DateTime.utc().toMillis() >= createdAtDateTimePlus60Minutes.toMillis()) {
+    return res.status(422).send({ error: 'This password reset token is invalid.' })
+  }
+
+  // Compare token
+  // const isTokenValid = await compare(token, passwordReset.token)
```
Comment (ericdowell) [inline src/controllers/auth/resetPassword.ts:95]: We'll want to use this line once we've updated to console logging the unhash token out to be used when calling this endpoint.

File: src/__tests__/auth/resetPassword.test.ts [hunk 7]
```diff
@@ -1,19 +1,88 @@
...
+            "param": "currentPassword",
+          },
+          Object {
+            "location": "body",
+            "msg": "Current Password is required.",
+            "param": "currentPassword",
+          },
+          Object {
+            "location": "body",
+            "msg": "data and hash arguments required",
```
Comment (ericdowell) [inline src/__tests__/auth/resetPassword.test.ts:28]: @larzhole I'm not sure this is the correct error message that we're expecting. Haven't looked into why this is return rather than `Current Password must match the saved password.`.

File: src/controllers/auth/resetPassword.ts (hunk 1 above)
Comment (ericdowell) [reply src/controllers/auth/resetPassword.ts:14]: I wasn't able to find something specific with Supertest, so I added JWT Magic Link login middleware. Also then fixed the tests, committed these changes in https://github.com/ericdowell/rerun/pull/11/commits/d8997c5bc8cc463cd0d32d3c5b429241a9ee90ea

File: src/__tests__/auth/resetPassword.test.ts (hunk 7 above)
Comment (ericdowell) [reply src/__tests__/auth/resetPassword.test.ts]: Fixed this in https://github.com/ericdowell/rerun/pull/11/commits/139bdabe496530eaa2f5965f62893ed88b27c486

File: src/migration/1593699407760-CreatePasswordResetsTable.ts (hunk 3 above)
Comment (larzhole) [reply src/migration/1593699407760-CreatePasswordResetsTable.ts:10]: @ericdowell I didn't think so either but I was not able to get it working without it... until today lol. I'll fix that

File: src/migration/1593699407760-CreatePasswordResetsTable.ts (hunk 2 above)
Comment (larzhole) [reply src/migration/1593699407760-CreatePasswordResetsTable.ts:17]: @ericdowell typeorm migration is not working for me anymore, I don't know how I broke it

```
$ npm run typeorm migration:run
> rerun@0.1.0 typeorm /Developer/Work/rerun
> node --require ts-node/register ./node_modules/typeorm/cli.js "migration:run"

query: SELECT * FROM `INFORMATION_SCHEMA`.`COLUMNS` WHERE `TABLE_SCHEMA` = 'rerun' AND `TABLE_NAME` = 'migrations'
query: SELECT * FROM `rerun`.`migrations` `migrations` ORDER BY `id` DESC
No migrations are pending
```

File: src/auth/passwordBroker.ts (hunk 5 above)
Comment (larzhole) [reply src/auth/passwordBroker.ts:9]: Gotcha, will do

File: src/controllers/auth/resetPassword.ts (hunk 1 above)
Comment (larzhole) [reply src/controllers/auth/resetPassword.ts:14]: Thanks!

File: src/controllers/auth/resetPassword.ts (hunk 4 above)
Comment (larzhole) [reply src/controllers/auth/resetPassword.ts:86]: @ericdowell No that's not working for me, it's saying it can't be parsed. There's gotta be a cleaner solution, but this was the best I could figure out after a few hours haha

```
invalid: Invalid
explanation: "the input "Fri Jul 10 2020 09:07:36 GMT-0500 (Central Daylight Time)" can't be parsed as ISO 8601"
reason: "unparsable"
```

Comment (ericdowell) [reply src/controllers/auth/resetPassword.ts:86]: Seems createdAt is a Date object, which Luxon has DateTime.fromJSDate()

Otherwise you can use the Date.getTime() method on the createdAt Date object to get the milliseconds which can be passed to a Luxon DateTime method.

File: src/migration/1593699407760-CreatePasswordResetsTable.ts (hunk 2 above)
Comment (ericdowell) [reply src/migration/1593699407760-CreatePasswordResetsTable.ts:15]: You can run:

npm run typeorm migration:revert

A few times until you’ve removed all migrations and then use migration:run again.

File: src/__tests__/__data__/passwordReset.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import { PasswordReset, User } from '../../entity'
+
+export const generatePasswordReset = async (attr?: {
+  user?: User
+  createdAt?: string
+  token?: string
+}): Promise<PasswordReset> => {
+  const passwordReset = new PasswordReset()
+  passwordReset.user = attr?.user ?? new User()
+  passwordReset.createdAt = attr?.createdAt ?? 'Wed Jul 08 2020 14:11:14 GMT-0500 (Central Daylight Time)'
```
Comment (ericdowell) [inline src/__tests__/__data__/passwordReset.ts:10]: I believe `createdAt` should be a `Date` object. Update this to use the following instead:
```ts
new Date('July 8, 2020 14:11:14')
```
Then you can update the type above to use `Date` instead of `string`.

File: src/controllers/auth/resetPassword.ts (hunk 4 above)
Comment (ericdowell) [reply src/controllers/auth/resetPassword.ts:86]: I can confirm that `createdAt` on `User` is a `Date` object via logging the following:
```ts
import { logger } from '../logger'

class User {
  toJSON(): any {
    logger.info(`createdAt ${this.createdAt.constructor.name}, ${this.createdAt.getTime()}`)
//...
  }
}
```

We can update the types based on this. See commit: https://github.com/ericdowell/rerun/pull/11/commits/ebd2295d4f89afcf35d90b9d15cfa1f3f2153e98

File: src/migration/1593699407760-CreatePasswordResetsTable.ts (hunk 2 above)
Comment (larzhole) [reply src/migration/1593699407760-CreatePasswordResetsTable.ts:15]: Thanks! That worked

File: src/controllers/auth/resetPassword.ts (hunk 4 above)
Comment (larzhole) [reply src/controllers/auth/resetPassword.ts]: I updated this now to be much cleaner using your suggestion

File: src/controllers/auth/resetPassword.ts (hunk 6 above)
Comment (larzhole) [reply src/controllers/auth/resetPassword.ts:94]: I don't remember what changed to cause that not to work

File: src/controllers/auth/resetPassword.ts [hunk 9]
```diff
@@ -83,17 +82,11 @@ export const postGuestResetPassword = async (req: Request, res: Response): Promi
     return res.status(404).json({ error: 'Token not found for the given email address' })
   }
 
-  const createdAt = passwordReset.createdAt.toString()
+  const expiration = DateTime.fromJSDate(passwordReset.createdAt, { zone: 'America/Chicago' })
```
Comment (ericdowell) [inline src/controllers/auth/resetPassword.ts:85]: This should not use `{ zone: 'America/Chicago' }` the database column is UTC.

File: src/controllers/auth/resetPassword.ts (hunk 6 above)
Comment (ericdowell) [reply src/controllers/auth/resetPassword.ts:94]: Well it's like I said, you are passing the hashed `token` database column instead of the actual `token`. So this will need to be updated to be commented back in and the next line removed.

File: src/controllers/auth/resetPassword.ts [hunk 10]
```diff
@@ -1,26 +1,107 @@
...
+  // Compare token
+  // const isTokenValid = await compare(token, passwordReset.token)
+  const isTokenValid = token === passwordReset.token
+
+  if (!isTokenValid) {
+    return res.status(500).json({ error: 'Invalid token' })
+  }
+
+  await PasswordReset.delete(passwordReset)
+  await User.update(user, { password: await hash(req.body.password, 10) })
```
Comment (ericdowell) [inline src/controllers/auth/resetPassword.ts:102]: We could place a try/catch around both of these lines and return `something went wrong` in the catch:
```suggestion
  try {
    await PasswordReset.delete(passwordReset)
    await User.update(user, { password: await hash(req.body.password, 10) })
  } catch(error) {
    logger.error(error)
    return res.status(500).json({ error: 'something went wrong' })
  }
```

File: src/controllers/auth/forgotPassword.ts [hunk 11]
```diff
@@ -12,11 +13,29 @@ export const postForgot = async (req: Request, res: Response): Promise<Response>
...
+
+  try {
+    await PasswordReset.save(passwordReset)
+  } catch (error) {
+    logger.error(error)
+    return res.status(500).json({ error: 'something went wrong' })
+  }
+
+  // TODO: email the token to the user/guest, return a 204 response
+  logger.info(passwordReset.token)
```
Comment (ericdowell) [inline src/controllers/auth/forgotPassword.ts:39]: You can update this to log `token`:
```suggestion
   logger.info(token) 
```
Since this is the correct value to pass to the guest reset password endpoint.

File: src/controllers/auth/forgotPassword.ts [hunk 12]
```diff
@@ -12,11 +13,29 @@ export const postForgot = async (req: Request, res: Response): Promise<Response>
...
-  // const payload = getPayload(token)
-  // can logger.info(token) for now, otherwise we want to email this to the user.
-  logger.info(token)
-  return res.status(501).json({ email, token })
+
+  let user: User
+
+  try {
+    const email = req.body.email
+    user = await User.findOneOrFail({ where: { email } })
```
Comment (ericdowell) [inline src/controllers/auth/forgotPassword.ts:21]: Can shorten this to not use a local variable `email`:
```suggestion
    user = await User.findOneOrFail({ where: { email: req.body.email } })
```

File: src/entity/PasswordReset.ts [hunk 13]
```diff
@@ -0,0 +1,21 @@
+import { BaseEntity, Column, Entity, JoinColumn, OneToOne, PrimaryColumn } from 'typeorm'
+import { hash } from 'bcrypt'
+import { User } from '../entity/User'
+
+@Entity({ name: 'passwordResets' })
+export class PasswordReset extends BaseEntity {
+  @PrimaryColumn({ name: 'userId', type: 'bigint', unsigned: true })
+  @OneToOne((type) => User, (user) => user.id)
```
[the remark this answers is not in the document]
Comment (ericdowell) [reply src/entity/PasswordReset.ts:8]: I also don't think we should use `@PrimaryColumn` and instead just use `@Column`, the ORM will attempt to increment if we use `@PrimaryColumn`

File: src/entity/PasswordReset.ts [hunk 14]
```diff
@@ -0,0 +1,21 @@
+import { BaseEntity, Column, Entity, JoinColumn, OneToOne, PrimaryColumn } from 'typeorm'
+import { hash } from 'bcrypt'
+import { User } from '../entity/User'
```
Comment (ericdowell) [inline src/entity/PasswordReset.ts:3]: Import can be shortened:
```suggestion
import { User } from './User'
```

File: src/entity/PasswordReset.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (ericdowell) [reply src/entity/PasswordReset.ts:8]: Although looking at the documentation it does say the following about [Creating a primary column](https://typeorm.io/#/undefined/creating-a-primary-column):
>Each entity **must** have at least one primary key column. This is a requirement and you can't avoid it. To make a column a primary key, you need to use `@PrimaryColumn` decorator.

Which I find to be complete bs... let's see if we can get away without using a primary key.

[the remark this answers is not in the document]
Comment (ericdowell) [reply src/entity/PasswordReset.ts:8]: Forget about replacing `@PrimaryColumn` with `@Column`, I get the following error:
```
 name: 'MissingPrimaryColumnError',
  message: 'Entity "PasswordReset" does not have a primary column. Primary column is required to have in all your entities. Use @PrimaryColumn decorator to add a primary column to your entity.'
```

File: src/controllers/auth/resetPassword.ts (hunk 6 above)
Comment (larzhole) [reply src/controllers/auth/resetPassword.ts]: Got it, yea I threw myself off by logging the hashed token like you said.
Fixed it now

File: src/controllers/auth/resetPassword.ts (hunk 9 above)
Comment (larzhole) [reply src/controllers/auth/resetPassword.ts]: Fixed

File: src/entity/PasswordReset.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (larzhole) [reply src/entity/PasswordReset.ts]: Fixed

File: src/controllers/auth/resetPassword.ts [hunk 15]
```diff
@@ -1,26 +1,107 @@
...
+  if (DateTime.utc().toMillis() >= expiration.toMillis()) {
+    return res.status(422).send({ error: 'This password reset token is invalid.' })
+  }
+
+  // Compare token
+  // const isTokenValid = await compare(token, passwordReset.token)
+  const isTokenValid = token === passwordReset.token
+
+  if (!isTokenValid) {
+    return res.status(500).json({ error: 'Invalid token' })
```
[the remark this answers is not in the document]
Comment (larzhole) [reply src/controllers/auth/resetPassword.ts]: Fixed

File: src/controllers/auth/resetPassword.ts (hunk 10 above)
Comment (larzhole) [reply src/controllers/auth/resetPassword.ts]: Fixed

File: src/controllers/auth/forgotPassword.ts (hunk 11 above)
Comment (larzhole) [reply src/controllers/auth/forgotPassword.ts]: Fixed

File: src/controllers/auth/forgotPassword.ts (hunk 12 above)
Comment (larzhole) [reply src/controllers/auth/forgotPassword.ts]: Fixed

File: src/entity/PasswordReset.ts (hunk 14 above)
Comment (larzhole) [reply src/entity/PasswordReset.ts]: Fixed

File: src/controllers/auth/resetPassword.ts [hunk 16]
```diff
@@ -1,26 +1,110 @@
...
     return res.status(422).json({ errors: errors.array() })
   }
-  return res.status(501).send()
+
+  try {
+    await User.update(req.user, { password: await hash(req.body.password, 10) })
+    return res.status(204).send()
+  } catch (error) {
+    logger.error(`Unable to update password for user with id: ${req.user.id}`)
+    return res.status(400).json({ error: 'User not found' })
```
Comment (ericdowell) [inline src/controllers/auth/resetPassword.ts:48]: We can return a 500, something went wrong here. We don’t know for sure that the user was not found.

Comment (larzhole) [reply src/controllers/auth/resetPassword.ts]: Fixed
